contracts/staking/ValidatorManager.sol

```diff
@@ -93,6 +90,105 @@ abstract contract ValidatorManager is
         $._pChainBlockchainID = settings.pChainBlockchainID;
         $._subnetID = settings.subnetID;
         $._maximumHourlyChurn = settings.maximumHourlyChurn;
+        $._initializedValidatorSet = false;
+    }
+
+    modifier initializedValidatorSet() {
+        require(
+            _getValidatorManagerStorage()._initializedValidatorSet,
+            "ValidatorManager: validator set not initialized"
+        );
+        _;
+    }
+
+    /**
+     * @notice See {IValidatorManager-initializeValidatorSet}.
+     */
+    function initializeValidatorSet(
+        SubnetConversionData calldata subnetConversionData,
+        uint32 messageIndex
+    ) external {
+        ValidatorManagerStorage storage $ = _getValidatorManagerStorage();
+        require(!$._initializedValidatorSet, "ValidatorManager: already initialized validator set");
+        require(
+            subnetConversionData.blockchainID == WARP_MESSENGER.getBlockchainID(),
+            "ValidatorManager: invalid blockchain ID"
+        );
+        require(
+            address(subnetConversionData.validatorManagerAddress) == address(this),
+            "ValidatorManager: invalid validator manager address"
+        );
+        require(
+            subnetConversionData.initialValidators.length > 0,
+            "ValidatorManager: no initial validators"
+        );
+        uint256 numInitialValidators = subnetConversionData.initialValidators.length;
+        require(
+            numInitialValidators < type(uint32).max, "ValidatorManager: too many initial validators"
+        );
+
+        // Verify that the sha256 hash of the Subnet conversion data matches with the Warp message's subnetConversionID.
+        bytes memory encodedConversion = abi.encodePacked(
+            subnetConversionData.convertSubnetTxID,
+            subnetConversionData.blockchainID,
+            uint32(20),
+            subnetConversionData.validatorManagerAddress,
+            uint32(numInitialValidators)
+        );
+        for (uint256 i; i < numInitialValidators; i++) {
```

Comment:
```suggestion
        for (uint256 i; i < numInitialValidators; ++i) {
```

Nit: Pre-increment is generally slightly more gas efficient for for loops since it results in one few store operation. Not sure if we've gone with post-increment in the past for any specific reason though.

Reply:
ye makes sense, I think we just forgot at times previously